src/main/java/org/radarcns/biovotion/BiovotionDeviceManager.java

```diff
@@ -317,6 +408,206 @@ public void onScanError(@NonNull Scanner scanner, @NonNull BleScanException thro
     }
 
 
+
+    /*
+     * ParameterListener interface
+     */
+
+    @Override
+    public void onParameterWritten(@NonNull final ParameterController ctrl, int id) {
+        logger.info("Biovotion VSM Parameter written: {}", id);
+
+        if (id == VsmConstants.PID_UTC) paramReadRequest(VsmConstants.PID_UTC);
+    }
+
+    @Override
+    public void onParameterWriteError(@NonNull final ParameterController ctrl, int id, final int errorCode) {
+        logger.error("Biovotion VSM Parameter write error, id={}, error={}", id, errorCode);
+    }
+
+    @Override
+    public void onParameterRead(@NonNull final ParameterController ctrl, @NonNull Parameter p) {
+        logger.info("Biovotion VSM Parameter read: {}", p);
+
+        // read algo_mode parameter; if not mixed algo/raw, set it to that
+        if (p.id() == VsmConstants.PID_ALGO_MODE) {
+            if (p.value()[0] != VsmConstants.MOD_MIXED_VITAL_RAW) {
+                // Set the device into mixed (algo + raw) mode. THIS WILL REBOOT THE DEVICE!
+                logger.warn("Biovotion VSM setting algo mode to MIXED_VITAL_RAW. The device will reboot!");
+                final Parameter algo_mode = Parameter.fromBytes(VsmConstants.PID_ALGO_MODE, new byte[] {(byte) VsmConstants.MOD_MIXED_VITAL_RAW});
+                paramWriteRequest(algo_mode);
+                //Boast.makeText(context, "Rebooting Biovotion device (switch to mixed algo mode)", Toast.LENGTH_LONG).show();
+            }
+        }
+
+        // read gsr_on parameter; if not on, activate
+        else if (p.id() == VsmConstants.PID_GSR_ON) {
+            if (p.value()[0] != 0x01) {
+                // Turn on GSR module
+                logger.info("Biovotion VSM activating GSR module");
+                final Parameter gsr_on = Parameter.fromBytes(VsmConstants.PID_GSR_ON, new byte[] {(byte) 0x01});
+                paramWriteRequest(gsr_on);
+            }
+        }
+
+        // read device UTC time
+        else if (p.id() == VsmConstants.PID_UTC) {
+            Date date = new Date(p.valueAsInteger() * 1000L);
+            SimpleDateFormat datef = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss z");
+            logger.info("Biovotion VSM device UTC timestamp is: {} ({})", p.valueAsInteger(), datef.format(date));
+            //Boast.makeText(context, "Biovotion device UTC: " + p.valueAsInteger(), Toast.LENGTH_LONG).show();
```

Comment:
Good one! I actually meant something else though: the result of this calculation does not seem to be stored anywhere, is that intentional? Alternative without a calendar is to just set
```java
Date date = new Date(p.valueAsInteger() * 1000L);
datef.setTimeZone(TimeZone.getTimeZone("UTC"));
datef.format(date);
```